Database service integration

Itxaka opened the pull request:
Missing:
- [x] ~~Spinner~~
- [x] ~~Tests~~
- [x] ~~Error handling~~ (out of scope)
- [x] ~~Running/complete states set~~
- [x] ~~Long running function to check database upgrade status after start~~
- [x] ~~Button disable on start~~

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -69,9 +75,20 @@
                 )
         }
+
+        function getErrorFromResponse(response) {
```
Missing docblock

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,92 @@
+              });
+
+            if (emptyFields) {
+                return
```
semicolon

Itxaka replied:
done

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,92 @@
+        // functions
+
+        function createServer() {
+            var emptyFields = _.some(
```
This looks like error handling implementation. We normally have one single card for it, but on this particular page, we have 2:
https://trello.com/c/w4ZiBPUh/30-3-16-create-connect-crowbar-db-ui-story-2-page-and-navigation
https://trello.com/c/rJcjpRVv/39-2-15-create-connect-crowbar-db-error-handling

Itxaka replied:
Indeed, this may be out of scope. I added it so it could easily be tested, but Im guessing that now that the ng-disabled directive is in place, its not needed anymore!

Itxaka replied:
oh wait, ng-disabled has nothing to do with it, this was to protect against empty fields when clicking the button and indeed, that should be dealt separately I guess? Althougth the form.isValid should enter in the scope for the card...button should be disabled until the fields are valid.

Thoughts? Im gonna leave it for the moment as its easy for testing, but is the form validation on the card scope?

santiph replied:
Leave it as it is if that makes your life and development easier. :) Just know that this will be reviewed/adjusted when we work on the other cards.

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,92 @@
+                )
+                .finally(function () {
+                    vm.running = false;
+                })
```
semicolon

Itxaka replied:
done

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,92 @@
+
+        function connectServer() {
+            var emptyFields = _.some(vm.databaseForm, function (x) {
+                return x == '' || angular.isUndefined(x)
```
`_.isUndefined(x) || _.isEmpty(x);`
Also, I'd change the variable name to `element` or `field` instead of `x`.

Itxaka replied:
ah, thats great and much nicer to read! done!

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.spec.js:
```diff
@@ -64,14 +94,124 @@ describe('Upgrade Flow - Create Connect Database Controller', function () {
+            });
+
+            it('errors should be empty', function () {
+                expect(controller.errors).toEqual([]);
```
`expect(controller.errors).to.be.empty;`
http://chaijs.com/api/bdd/#method_empty

Itxaka replied:
weird. With this I get `undefined is not an object (evaluating 'expect(controller.errors).to.be')` but if I dump the object I get the proper `[]` ¯\_(ツ)_/¯

skazi0 replied:
@santiph this will not work as jasmine is overriding chai's `expect()`. To make this work you would need to locally switch to chai e.g. `var expect = window.chai.expect;` or call it directly via `chai.expect(...).to.be.empty;`.

santiph replied:
Good point. @Itxaka I'd use `chai.expect` then

Itxaka replied:
done

santiph reviewed (changes requested):
A couple of questions and changes to be addressed

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [#14 Create/Connect Crowbar DB: Service Integration](https://trello.com/c/tslrmrBm/157-14-create-connect-crowbar-db-service-integration)

Commit:
Database: Service Integration

 - Database migration for new server ot connecting to an existing server
 - Remove uneeded field (Table Prefix)
 - Tests for new functionality
 - Fake API calls that mimic the real ones, with a 4 second delay on the response
 - Buttons are disabled if the fields are empty or the database calls is running or if the database finished

Itxaka commented:
@santiph @skazi0 Ready for review

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,75 @@
+        function createServer() {
+            vm.errors = [];
+            vm.running = true;
+            vm.spinnerVisible = true;
```
`spinnerVisible` should not be changed by the controller. It's a 'feedback' variable used by the spinner directive to indicate that the spinner is actually visible.

Itxaka replied:
Indeed!

skazi0 commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,16 +11,75 @@
+        function connectServer() {
+            vm.errors = [];
+            vm.running = true;
+            vm.spinnerVisible = true;
```
ditto

skazi0 commented on assets/app/features/upgrade/i18n/en.json:
```diff
@@ -61,19 +61,19 @@
                 "navigation-tabs": {
-                    "create-database": "Create New Database",
-                    "connect-database": "Connect to an Existing Database"
+                    "create-server": "Create New Server",
```
Does it really create new _server_?

Itxaka replied:
Yup!
It will run a chef-solo client to install and configure the postgres server on the administration server.

Still this will change in the future as @kwwii said that all text and so on needs to be refined, so I wouldnt give it much thougth rigth now.

Itxaka replied:
well, not a _server_ per se but a _database server_ :+1:

santiph replied:
We need to align to Ken's copys as specified in: https://github.com/crowbar/crowbar-ui/wiki/Upgrade67-Create-or-Connect-To-PostgreSQL-Database

Itxaka replied:
Im guessing the refinement of this will come later, because there is a huge difference in what it says and in what it _does_.

santiph replied:
Until further specifications are required, we're supposed to implement what has been already defined in the comps

Commit:
- Do not set the spinner visibility by the controller

Commit:
- Reverse naming change
 - Use chai.expect for some tests

skazi0 reviewed (approved):
lgtm

skazi0 committed:
upgrade: Fixed base functionality of db configuration page.

skazi0 committed:
upgrade: Database form validation

Itxaka commented on gulpfile.js:
```diff
@@ -41,6 +41,7 @@ gulp.task('default', function (callback) {
         'build',
         'rsync',
         'watch',
+        'tdd',
```
Mixing PR purpouse with extra stuff? Should be separated to a different PR (probably a slip?)

skazi0 replied:
yeah, sorry for that. reverted.

Itxaka commented:
@santiph please re-review :D

santiph commented on assets/app/data/crowbar/services/upgrade.factory.js:
```diff
@@ -81,7 +83,36 @@
+        }
+
+        /**
+        * Install, configure and migrate a new database server on the administration server
```
Migrate a new database server? It doesn't makes much sense...

Itxaka replied:
lmao, indeed. Not sure what was I thinking when writing that, it makes no sense

santiph commented on assets/app/data/crowbar/services/upgrade.factory.js:
```diff
@@ -81,7 +83,36 @@
+        }
+
+        /**
+        * Configure and migrato to an existing database server
```
typo: `migrato` -> `migrate`

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,20 +11,86 @@
-        upgradeStepsFactory.setCurrentStepCompleted();
 
+        /* eslint-disable max-len */
+        // NOTE: patterns based on https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/mixin/database.rb
```
Thanks for this!

santiph commented on assets/app/features/upgrade/controllers/upgrade-database-configuration.controller.js:
```diff
@@ -11,20 +11,86 @@
+        vm.patterns.host = vm.patterns.hostname + '|' + vm.patterns.ipv4;
+
+        vm.running = false;
+        vm.spinnerVisible = false;
```
@skazi0 Isn't this the attribute intended to be a notification only?

skazi0 replied:
Not sure I understand your question. If you mean that this one doesn't need to be defined in the controller then yes, it can be done implicitly by angular.

santiph commented on routes/api/upgrade/connect.js:
```diff
@@ -0,0 +1,70 @@
+var express = require('express'),
+    router = express.Router();
+
+var responseOk = {
```
Completely minor, specially for a mocked service like this: You can use the same `var` statement in line #1 like this:
```
var express = require('express'),
    router = express.Router(),
    responseOk = {
```

Itxaka replied:
woops, will fix

santiph commented on routes/api/upgrade/connect.js:
```diff
@@ -0,0 +1,70 @@
+            'success': true
+        },
+        'crowbar_init': {
+            'success': true
```
If the database setup step failed, shouldn't the database migration, schema migration and crowbar init tasks also not be successful?

Itxaka replied:
I think I copied the response directly from the apiDocs so that may be the reason its like that? but indeed, all of them should be changed to failed

skazi0 replied:
The real question is: what will the backend return in such cases?

Itxaka replied:
Updated with the *actual* response provided in the apiDocs. Not sure where did I get this one from...

santiph reviewed (changes requested):
Partially reviewed, I'm still missing the view implementation (Will check it later today after some meetings).
Some changes requested.

Itxaka commented:
rebased

santiph commented on assets/app/features/upgrade/templates/database-configuration-page.jade:
```diff
@@ -1,45 +1,83 @@
-                        .form-group
-                            label(translate='') upgrade.steps.database-configuration.database-form.username
-                            input.form-control(type='text', name='username', ng-required='false', ng-model='upgradeDatabaseVm.databaseForm.username')
+                        .form-group(ng-class="{'has-error': newForm.username.$invalid && newForm.username.$dirty}")
```
Is `has-error` actually needed, considering we already have `ng-invalid`, `ng-dirty` and `ng-touched`?
@see: https://docs.angularjs.org/guide/forms#using-css-classes

skazi0 replied:
This is binding Angular's validation states with Bootstrap's classes: http://getbootstrap.com/css/#forms-control-validation

skazi0 replied:
As an alternative, we could consider using something like: https://www.npmjs.com/package/angular-bootstrap-form-validation. It handles `has-error` internally and in addition provides hints on the cause of validation failure.

Note that this one is not usable directly as it doesn't currently support `ngPattern` and has some old dependencies in `bower.json`. But maybe there are some newer alternatives with similar functionality. To be considered.

santiph commented on assets/app/features/upgrade/templates/database-configuration-page.jade:
```diff
@@ -1,45 +1,83 @@
-                                input.form-control(type='text', required='', ng-model='upgradeDatabaseVm.databaseForm.port')
+                                label.control-label(for='port', translate='') upgrade.steps.database-configuration.database-form.port
+                                input.form-control(type='number', name='port',
+                                min='1', max='65535',
```
Are `1` and `65365` allowed or excluded?

Itxaka replied:
allowed

Commit:
upgrade: minor fixes for the db step

Itxaka commented:
Updated to address the comments. Tested but its not working, button wont fire any event and no errors are shown

Itxaka commented:
Disregard last comment, seems my cache had leftovers from olders patches. Cleaning and "gulping" again, everything works beautifully. Kudos to @skazi0 for the pattern matching, works wonderfully!

Itxaka committed:
upgrade: clarify db comment

santiph commented:
Looks good to me. I completely miss Bootstrap on the view you've implemented, thanks @skazi0 for remind me. `angular-bootstrap-form-validation` looks interesting, but I'd leave it as it is for now. We could analyse it and look for alternatives as part of the proper error handling cards we have.
